### PR TITLE
Pandas-only P2P shuffling

### PR DESCRIPTION
**Pull request by hendrikmakait**
Supersedes #8606

- [ ] Tests added / passed
- [ ] Passes `pre-commit run --all-files`

### CODE REVIEW

**Comment by rjzamora** [comment]
@hendrikmakait - Is there any background or motivation you can share about this work? Seems very pandas-specific at the moment.

File: distributed/shuffle/_pickle.py [hunk 1]
```diff
@@ -40,3 +45,96 @@ def unpickle_bytestream(b: bytes | bytearray | memoryview) -> Iterator[Any]:
...
+    # Actually load memory-mapped buffers into memory and close the file
+    # descriptors
+    return dd.methods.concat(shards, copy=True)
+
+
+def restore_dataframe_shard(
+    index: pd.Index, blocks: Sequence[Block], meta: pd.DataFrame
+) -> pd.DataFrame:
+    import pandas as pd
+    from pandas.core.internals import BlockManager, make_block
```
Comment (rjzamora) [inline distributed/shuffle/_pickle.py:114]: Would be nice in the long run if any logic using pandas internals could be dispatched.

File: continuous_integration/environment-mindeps.yaml [hunk 2]
```diff
@@ -22,7 +22,7 @@ dependencies:
   # Distributed depends on the latest version of Dask
   - pip
   - pip:
-    - git+https://github.com/dask/dask
+    - git+https://github.com/hendrikmakait/dask@p2p-pandas
```
Comment (hendrikmakait) [inline continuous_integration/environment-mindeps.yaml:25]: Reminder to revert https://github.com/dask/distributed/pull/8635/commits/814c73ad16a91a76c7b59b12d1752e5c0c6f1645

**Comment by mrocklin** [comment]
> Is there any background or motivation you can share about this work? Seems very pandas-specific at the moment.

I'll let @hendrikmakait answer fully, but my sense is that early experiments showed substantial speedups (like 30% on full TPC-H Benchmarks, not just on the shuffling portion).

> Would be nice in the long run if any logic using pandas internals could be dispatched.

I can see how this would be convenient for RAPIDS folks (although probably a little inconvenient for others due to complexity).  If we go this direction then maybe helping address/reduce this complexity is something that RAPIDS people could help design/execute/maintain?

**Comment by rjzamora** [comment]
>my sense is that early experiments showed substantial speedups (like 30% on full TPC-H Benchmarks, not just on the shuffling portion).

Very nice!

>I can see how this would be convenient for RAPIDS folks (although probably a little inconvenient for others due to complexity).

This is a breaking change for rapids "as is". I suppose it would be "convenient" for rapids to continue working with p2p shuffling :)

>If we go this direction then maybe helping address/reduce this complexity is something that RAPIDS people could help design/execute/maintain?

I will certainly help with all of the above if doing so is both welcome and feasible. Feasibility probably depends on it being possible to isolate the logic that interacts with pandas internals.  If organizing the code in this way really is "too complex", then cudf may need to rely on a completely distinct version of "p2p" to avoid slowing down Coiled engineers. However, my hope is that this is a natural way to organize things even if cudf wasn't in the picture. (fingers crossed)

For this PR, it seems fine to focus on pandas-only details (especially for now). @hendrikmakait - Perhaps we can meet offline so I can figure out whether rapids needs to prioritize (1) rolling-back p2p support, or (2) enabling the pyarrow-free approach with cudf-backed data.

**Comment by hendrikmakait** [comment]
> @hendrikmakait - Is there any background or motivation you can share about this work? Seems very pandas-specific at the moment.

There are three main benefits I see here:
1. As @mrocklin mentioned, we have seen _significant_ performance improvements with this. I plan to share some benchmarks later. (The Arrow-based implementation has never been optimized for performance and has some shortcomings like excessive data copying.)  
3. It removes the Arrow dependency, a consistent source of pain because of semantics differing from pandas and its strict typing.
4. It removes complexity that never worked as intended, e.g., disk-write buffering.

> For this PR, it seems fine to focus on pandas-only details (especially for now). @hendrikmakait - Perhaps we can meet offline so I can figure out whether rapids needs to prioritize (1) rolling-back p2p support, or (2) enabling the pyarrow-free approach with cudf-backed data.

Sounds good! I'll ping you offline. Looking at the code we currently have in place, I expect it to be fairly straightforward to dispatch the library-specific pieces of code or to dispatch instantiation of shuffle runs and implement a cuDF-specific version.

File: distributed/shuffle/_disk.py [hunk 3]
```diff
@@ -210,6 +226,27 @@ def read(self, id: str) -> Any:
...
+        all arrays. This is a fast sequence of short reads interleaved with seeks.
+        Do not read in memory the actual data; the arrays' buffers will point to the
+        memory-mapped area.
+
+        The file descriptor will be automatically closed by the kernel when all the
+        returned arrays are dereferenced, which will happen after the call to
+        concatenate3.
+        """
+        with path.open(mode="r+b") as fh:
+            buffer = memoryview(mmap.mmap(fh.fileno(), 0))
```
Comment (hendrikmakait) [inline distributed/shuffle/_disk.py:240]: The `mmap` causes us to lose strict control over the file descriptor lifecycle. This creates problems with cleanup on Windows because we can't remove files if they still have an open file descriptor.

File: distributed/shuffle/_shuffle.py [hunk 4]
```diff
@@ -310,81 +301,52 @@ def _construct_graph(self) -> _T_LowLevelGraph:
...
-    partitions.sort()
-    t = t.sort_by(column)
+    out: defaultdict[str, list[tuple[int, list[PickleBuffer]]]] = defaultdict(list)
 
-    partition = np.asarray(t[column])
-    if drop_column:
-        t = t.drop([column])
-    splits = np.where(partition[1:] != partition[:-1])[0] + 1
-    splits = np.concatenate([[0], splits])
+    base = df[column].values.base  # type: ignore[union-attr]
```
Comment (rjzamora) [inline distributed/shuffle/_shuffle.py:335]: Can you explain the `base` logic a bit?
In cudf/cupy, `df[column].values.base` produces `None`.

Does it make sense to protect against the case that `base in None` here? (Otherwise, `len(base)` will produce an error)

File: distributed/shuffle/_pickle.py [hunk 5]
```diff
@@ -40,3 +45,98 @@ def unpickle_bytestream(b: bytes | bytearray | memoryview) -> Iterator[Any]:
...
+) -> list[pickle.PickleBuffer]:
+    """Optimized pickler for pandas Dataframes. DIscard all unnecessary metadata
+    (like the columns header).
+
+    Parameters:
+        obj: pandas
+    """
+    return pickle_bytelist(
+        (input_part_id, shard.index, *shard._mgr.blocks), prelude=False
+    )
```
Comment (rjzamora) [inline distributed/shuffle/_pickle.py:62]: Fun fact: This PR works fine with cudf-backed data when I do something like the following:

```python
    shard = shard.to_pandas()  if hasattr(shard, "to_pandas") else shard
    return pickle_bytelist(
        (input_part_id, shard.index, *shard._mgr.blocks), prelude=False
    )
```

as long as I also modify the last line of `unpickle_and_concat_dataframe_shards` to be:

```python
    result = dd.methods.concat(shards, copy=True)
    if hasattr(type(meta), "from_pandas"):
        return type(meta).from_pandas(result)
    return result
```

I'm sure there are cases where the cudf->pandas->cudf round trip isn't perfect, but the small amount of logic I needed to work-around the changes is promising.

**Comment by rjzamora** [comment]
I've been experimenting with PR using cudf-backed data. I think it will take some work on the rapids side to bring performance in line with `main` (tpch queries take a ~2x perf hit with dask-cudf), but I wouldn't consider cudf + p2p performance to be a big priority yet. My near-term priority is to keep cudf + p2p "functional". This can be achieved in a fairly simple way by defining/using the following dispatch functions:

```python
def deconstruct_dataframe_shard(shard: pd.DataFrame) -> tuple[Any]:
    """ Deconstruct a DataFrame shard into the information needed to reconstruct it later

    Dispatch on the type of `shard`.
    The elements of the result should be whatever is needed by `restore_dataframe_shard`
    """
    ...

def restore_dataframe_shard(meta: pd.DataFrame, unpickled_shard: tuple[Any]) -> pd.DataFrame:
    """ Reconstruct an un-pickled DataFrame shard

    Dispatch on the type of `meta`.
    `unpickled_shard` corresponds to the round-tripped output of `deconstruct_dataframe_shard`.
    Converts a tuple of un-pickled data (e.g. index  and blocks) back to a DataFrame.
    """
    ...
```

For the sake of generality, `pickle_dataframe_shard` would look something like:

```python
def pickle_dataframe_shard(input_part_id: int, shard: pd.DataFrame) -> list[pickle.PickleBuffer]:
    return pickle_bytelist(
        (input_part_id,) +  deconstruct_dataframe_shard(shard),
        prelude=False,
    )
```

and `unpickle_and_concat_dataframe_shards` would then look something like:

```python
def unpickle_and_concat_dataframe_shards(
    b: bytes | bytearray | memoryview, meta: pd.DataFrame
) -> pd.DataFrame:
    import dask.dataframe as dd

    unpickled_shards = list(unpickle_bytestream(b))
    unpickled_shards.sort(key=first)
    shards = []

    for _, unpickled_shard in unpickled_shards:
        shards.append(restore_dataframe_shard(meta, unpickled_shard))

    return dd.methods.concat(shards, copy=True)
```

I suppose the only "complex" part of this general suggestion is that the dispatch functions themselves would need to be exposed/defined in `dask.dataframe`.

**Comment by hendrikmakait** [comment]
**TL;DR:** After running several A/B tests and profiling the code, I have concluded not to move forward with this PR because it is not a universal improvement.


**Analysis**

A/B tests have shown improvements ranging from 5%-30% in _end-to-end_ runtime for almost all TPC-H queries:
Screenshot 2024-05-14 at 16 49 45

However, they have also shown disastrous results for [`test_set_index`](https://github.com/coiled/benchmarks/blob/934a69e0ed093ef7319a5034b87c03a53dc0c0d8/tests/benchmarks/test_join.py#L69-L85) and to a lesser extent, [`test_set_index_uber_lyft`](https://github.com/coiled/benchmarks/blob/934a69e0ed093ef7319a5034b87c03a53dc0c0d8/tests/benchmarks/test_join.py#L89-L94).

Screenshot 2024-05-14 at 16 47 51



It looks like this PR is overfitting on OLAP-style queries that include narrow dataframes with few columns and heavily reduce data at the cost of ETL-style transformations that include wide dataframes with many columns.

Profiling the code before and after, I saw that both the transfer and the `unpack` phase have become significantly more expensive in `test_set_index` (by a factor of 3.5x and 2.4x, respectively) and we lose much time due to what looks like GIL contention. The Coiled clusters I ran these on also report a GIL contention very close to 1 (and higher than on `main`). 

This is likely due to many small shards with very few rows. I have tried multiple possible quick wins like removing `mmap` to avoid the impact of page loads during GIL-restricted code paths or using `df.drop()` instead of `del`, but none of these had any positive impact. It's currently unclear how to improve this implementation in a way that reduces GIL contention such that ETL-style workloads are not impact as harshly. There may be upstream improvements we could make that reduce GIL contention but those would also take a while to get released. I also suspect that this is yet another example of the convoy effect, so reducing GIL contention in some parts will have non-trivial effects as it had in previous attempts at improving P2P performance.

**Comment by hendrikmakait** [comment]
Example py-spy profiles for future reference:

[profiles.zip](https://github.com/dask/distributed/files/15406617/profiles.zip)
